src/Compilers/Core/Portable/SourceGeneration/IncrementalValueSource.cs

```diff
@@ -20,10 +20,12 @@ namespace Microsoft.CodeAnalysis
     public readonly struct IncrementalValueSource<T>
     {
         internal readonly IIncrementalGeneratorNode<T> Node;
+        internal readonly Action<IIncrementalGeneratorOutputNode> RegisterOutput;
 
-        internal IncrementalValueSource(IIncrementalGeneratorNode<T> node)
+        internal IncrementalValueSource(IIncrementalGeneratorNode<T> node, Action<IIncrementalGeneratorOutputNode> registerOutput)
         {
             this.Node = node;
+            this.RegisterOutput = node is SyntaxInputNode<T> sin ? sin.RegisterOutputNode : registerOutput;
```

Comment:
This is the one bit i'm not sure about. It works, but seems weird to special case and then ignore the input.

Reply:
Why is this special case needed?

Reply:
Should we have a different constructor here that takes `SyntaxInputNode` directly and assert in this ctor that it's not `SyntaxInputNode`? This pattern does seem a bit odd because in the one call site I see where we use `SyntaxInputNode` both `sin.RegisterOutputNode` and `registerOutput` are identical. Think it would be cleaner + clearer if we had separate ctors.

Reply:
Yeah, the reason for this is because we want to replace the registration function in this case, rather than using the existing one.

Consider:
- Start with a single syntax input `S`
- Node has a reference to `S.RegisterOutput`
- User calls `WithTransform` which creates syntax input `S'`
- Above code replaces `_registerOutput` with `S'.RegisterOutput`
- User calls GeneratorSource which invokes `_registerOutput` 
- `S'` gets registered.

We want `_registerOuput` to call the instance on `S'`. If we pass in the registration function at the original callsite only, then `_registerOutput` is still `S.RegisterOutput` and `S` gets registered instead of `S'`.

I considered making the speical case be an interface or something, so the node code can say 'use the existing one unless the node is `ISupportsRegistration`. The purest in me likes it, the pragmatist says we'll probably never need it, and we can make it when we do.

This is possible a deficiency in the node design tbh, there is no way to actually track back through them linked list style. If there was we could just do that at registration and go up the chain to register the input.

Reply:
> Node has a reference to S.RegisterOutput

In this case though where is `S.RegisterOutput` different that `_registerOutput`? The case I see which seems to fit this has them as both the same value.